### PR TITLE
[New rule] Add strict-id-in-types rule

### PR DESCRIPTION
**Pull request by eddeee888**
## Description

This PR adds a new rule `strict-id-in-types`. 

Usually, a GraphQL output type can be mapped to a domain model e.g. `User`, `Post`, `Comment`. These fields should have a logical database `id` which could be used as the unique identifier of the correspondent GraphQL output type. There are some exceptions: Payload/Result, Pagination, or types used to group related fields together.

This rule enforces all output types to have a unique identifier field unless belonging to the exceptions. This ensures the consistency of output types in the schema.

The default name is `id` and the default type is `ID`. A lot of GraphQL clients e.g. Apollo client and urql uses the `id` field of incoming data to normalise the cache so consistency from the schema level will help writing documents as well.

### Options

- `acceptedIdNames`: an array of strings containing field names that could be used as unique identifier. Works in tandem with `acceptedIdTypes`. Default: `["id"]`
- `acceptedIdTypes`: an array of strings containing field types that could be used as unique identifier. Works in tandem with `acceptedIdNames`. Default: `["ID"]`
- `exceptions`: an object containing properties of a type to be ignored.
  - `suffixes`: Any type that has one of the options declared here should be ignored. Default: `[]`

Related https://github.com/dotansimha/graphql-eslint/issues/326

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

N/A

## How Has This Been Tested?

I have added the tests in this file: https://github.com/eddeee888/graphql-eslint/blob/2f7efe34967cf26711897bcc614461eb3aa788c0/packages/plugin/tests/strict-id-in-types.spec.ts
To test, run `yarn test strict-id-in-types`

**Test Environment**:
- OS: MacOS
- `@graphql-eslint/...`: Latest versions
- NodeJS: 12.13.1

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests and linter rules pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

- When running `yarn generate:docs`, it changed a lot of unrelated doc files. This makes me think that the latest version on `master` may not have run generate:docs? 🤔  Please let me know how we should deal with this issue.
- It would be good to try this on an alpha version for some of my projects before releasing. Please let me know if it's possible :)

### CODE REVIEW

File: docs/rules/provided-required-arguments.md [hunk 1]
```diff
@@ -7,4 +7,4 @@
 
 A field or directive is only valid if all required (non-null without a default value) field arguments have been provided.
 
-> This rule is a wrapper around a `graphql-js` validation function. [You can find it's source code here](https://github.com/graphql/graphql-js/blob/master/src/validation/rules/ProvidedRequiredArgumentsRule.js).
\ No newline at end of file
+> This rule is a wrapper around a `graphql-js` validation function. [You can find it's source code here](https://github.com/graphql/graphql-js/blob/master/src/validation/rules/ProvidedRequiredArguments.js).
```
Comment (eddeee888) [inline docs/rules/provided-required-arguments.md:10]: A lot of these changes are unrelated to my change. Probably the version on `master` has not generated docs? 🤔

File: packages/plugin/src/rules/strict-id-in-types.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        const validIds = node.fields.filter(field => {
+          const fieldNode = field.rawNode();
+
+          const isValidIdName = options.acceptedIdNames.includes(fieldNode.name.value);
+
+          // To be a valid type, it must be non-null and one of the accepted types.
+          let isValidIdType = false;
+          if (fieldNode.type.kind === 'NonNullType' && fieldNode.type.type.kind === 'NamedType') {
+            isValidIdType = options.acceptedIdTypes.includes(fieldNode.type.type.name.value);
+          }
```
Comment (eddeee888) [inline packages/plugin/src/rules/strict-id-in-types.ts:147]: Is there a better way to check node types using some kind of helper from the `graphql` package? I tried so many approaches but couldn't get this working so I went with this. :)

I'm hoping to get some advice/guidance here 🙏

File: packages/plugin/src/rules/strict-id-in-types.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        }
+
+        const validIds = node.fields.filter(field => {
+          const fieldNode = field.rawNode();
+
+          const isValidIdName = options.acceptedIdNames.includes(fieldNode.name.value);
+
+          // To be a valid type, it must be non-null and one of the accepted types.
+          let isValidIdType = false;
+          if (fieldNode.type.kind === 'NonNullType' && fieldNode.type.type.kind === 'NamedType') {
```
Comment (eddeee888) [inline packages/plugin/src/rules/strict-id-in-types.ts:145]: Is there a better way to check node types using some kind of helper from the `graphql` package? I tried so many approaches but couldn't get this working so I went with this. :)

I'm hoping to get some advice/guidance here 🙏

File: docs/rules/provided-required-arguments.md (hunk 1 above)
Comment (dotansimha) [reply docs/rules/provided-required-arguments.md:10]: Yeah, maybe it's not fully updated. just ignore it for now :)

File: packages/plugin/src/rules/strict-id-in-types.ts (hunk 3 above)
Comment (dotansimha) [reply packages/plugin/src/rules/strict-id-in-types.ts:145]: `graphql.js` has `isNonNullType` but it's not for AST, it's only for class types. 

I'm totally fine with the way you tested it, but maybe you can use `Kind.NON_NULL_TYPE` and `Kind.NAMED_TYPE` instead of the explicit string valus?

**Review by dotansimha — APPROVED** [review]
Looks great, thank you @eddeee888 ! Please see my minor comment, other than that, it looks perfect!

**Comment by dotansimha** [comment]
Oh and one more thing: Maybe we can add additional fields for `exceptions`, something like `types:` to specify exclusions with the complete type name?

File: packages/plugin/src/rules/strict-id-in-types.ts (hunk 3 above)
Comment (eddeee888) [reply packages/plugin/src/rules/strict-id-in-types.ts]: Makes sense! This is done in https://github.com/dotansimha/graphql-eslint/pull/333/commits/6ba418a98de2823f98a605e117f034fb12090d39. Cheers!

**Comment by eddeee888** [comment]
Good idea @dotansimha ! 👍 

`exceptions.types` is implemented in https://github.com/dotansimha/graphql-eslint/pull/333/commits/4c79a4a385f50f27c7894bce94a015e0303178fa. Please take a look when you have a chance :)
